File: onebusaway-android/src/main/java/org/onebusaway/android/speed/VehicleTrajectoryTracker.java
/*
 * Copyright (C) 2024 Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.speed;

import org.onebusaway.android.io.elements.ObaRoute;

import java.util.HashMap;
import java.util.Map;

/**
 * Singleton speed-estimation facade. Delegates trip data access to
 * {@link TripDataManager} and owns only speed estimation logic
 * (gamma model, schedule speed, route type routing).
 */
public final class VehicleTrajectoryTracker {

    private static final VehicleTrajectoryTracker INSTANCE = new VehicleTrajectoryTracker();

    private final TripDataManager dataManager = TripDataManager.getInstance();
    private final Map<String, Integer> routeTypeCache = new HashMap<>();
    private final ScheduleSpeedEstimator scheduleEstimator = new ScheduleSpeedEstimator();
    private SpeedEstimator estimator = new GammaSpeedEstimator();

    private VehicleTrajectoryTracker() {
    }

    public static VehicleTrajectoryTracker getInstance() {
        return INSTANCE;
    }

    /**
     * Returns the estimated speed in m/s for the given key and current state.
     */
    public synchronized Double getEstimatedSpeed(String key, VehicleState state) {
        if (key == null || state == null) {
            return null;
        }
        Integer routeType = routeTypeCache.get(key);
        SpeedEstimator est = (routeType != null && ObaRoute.isGradeSeparated(routeType))
                ? scheduleEstimator : estimator;
        return est.estimateSpeed(state.getVehicleId(), state, dataManager);
    }

    /**
     * Returns the estimated speed in m/s for the given key, using the last cached VehicleState.
     */
    public synchronized Double getEstimatedSpeed(String key) {
        return getEstimatedSpeed(key, dataManager.getLastState(key));
    }

    /**
     * Returns the schedule-derived speed from the last speed estimate.
     */
    public synchronized double getLastScheduleSpeed() {
        return estimator.getLastScheduleSpeed();
    }

    /**
     * Returns the GammaParams from the last speed estimate.
     */
    public synchronized GammaSpeedModel.GammaParams getLastGammaParams() {
        return estimator.getLastGammaParams();
    }

    /**
     * Sets the active speed estimator.
     */
    public synchronized void setEstimator(SpeedEstimator estimator) {
        if (estimator != null) {
            this.estimator = estimator;
        }
    }

    /**
     * Stores the route type for a trip ID.
     */
    public synchronized void putRouteType(String tripId, int type) {
        if (tripId != null) {
            routeTypeCache.put(tripId, type);
        }
    }

    /**
     * Returns the cached route type for the given trip, or null if not cached.
     */
    public synchronized Integer getRouteType(String tripId) {
        return routeTypeCache.get(tripId);
    }

    /**
     * Clears speed estimation state and route type cache.
     */
    public synchronized void clearAll() {
        routeTypeCache.clear();
        estimator.clearState();
        scheduleEstimator.clearState();
    }
}

File: onebusaway-android/src/main/java/org/onebusaway/android/extrapolation/VehicleTrajectoryTracker.kt
/*
 * Copyright (C) 2024-2026 Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.extrapolation

import android.location.Location
import org.onebusaway.android.extrapolation.data.TripDataManager
import org.onebusaway.android.extrapolation.math.prob.ProbDistribution
import org.onebusaway.android.io.elements.ObaRoute
import org.onebusaway.android.util.LocationUtils

/**
 * Singleton extrapolation facade. Manages per-trip [Extrapolator] instances,
 * creating the appropriate type based on route type (gamma for buses,
 * schedule replay for grade-separated transit).
 */
object VehicleTrajectoryTracker {

    /** Max age of the newest AVL entry before we consider extrapolation unreliable. */
    const val MAX_EXTRAPOLATION_AGE_MS = 5L * 60 * 1000

    @JvmStatic fun getInstance() = this

    private val dataManager = TripDataManager
    private val extrapolators = HashMap<String, Extrapolator>()

    private fun getOrCreateExtrapolator(tripId: String): Extrapolator =
            extrapolators.getOrPut(tripId) {
                val routeType = dataManager.getRouteType(tripId)
                if (routeType != null && ObaRoute.isGradeSeparated(routeType))
                    ScheduleReplayExtrapolator(tripId, dataManager)
                else
                    GammaExtrapolator(tripId, dataManager)
            }

    /**
     * Extrapolates a distribution over distance along the trip at [queryTimeMs].
     * Returns null if extrapolation is not possible.
     */
    @Synchronized
    fun extrapolate(tripId: String, queryTimeMs: Long): ProbDistribution? =
            getOrCreateExtrapolator(tripId).extrapolate(queryTimeMs)

    /**
     * Extrapolates a vehicle's lat/lng position, writing into [out].
     * Returns true if a valid position was computed.
     */
    @Synchronized
    fun extrapolatePosition(tripId: String, queryTimeMs: Long, out: Location): Boolean {
        val sd = dataManager.getShapeWithDistances(tripId) ?: return false
        if (sd.points.isEmpty()) return false
        val dist = extrapolate(tripId, queryTimeMs) ?: return false
        return LocationUtils.interpolateAlongPolyline(
                sd.points, sd.cumulativeDistances, dist.median(), out)
    }

    /** Returns true if the tracker has recent enough AVL data to extrapolate this trip. */
    fun canExtrapolate(tripId: String, queryTimeMs: Long): Boolean {
        val last = dataManager.getLastState(tripId) ?: return false
        return queryTimeMs - last.lastLocationUpdateTime <= MAX_EXTRAPOLATION_AGE_MS
    }

    /** Clears all per-trip extrapolator instances. */
    @Synchronized
    fun clearAll() {
        extrapolators.clear()
    }
}
